src/utils/jwt.ts

```diff
@@ -0,0 +1,240 @@
+        const resp = err.response;
+        if (resp.isJson() && resp.data.error) {
+          errorMessage += `${resp.data.error}`;
+          if (resp.data.error_description) {
+            errorMessage += ' (' + resp.data.error_description + ')';
+          }
+        } else {
+          errorMessage += `${resp.text}`;
+        }
+        throw new JwtError(JwtErrorCode.INTERNAL_ERROR, errorMessage);
+      }
+      throw err;
+    });
+  }
+}
+
+/**
+ * Verifies JWT signature with a public key.
+ */
+export class PublicKeySignatureVerifier implements SignatureVerifier {
+  constructor(private keyFetcher: UrlKeyFetcher) {
+    if (!validator.isNonNullObject(keyFetcher)) {
+      throw new Error('The provided key fetcher is not an object or null.');
+    }
+  }
+
+  public verify(token: string): Promise<void> {
+    const ALGORITHM_RS256: jwt.Algorithm = 'RS256' as const;
+    const error = new JwtError(JwtErrorCode.INVALID_TOKEN,
+      'The provided token has invalid signature.');
+    return isSignatureValid(token, getKeyCallback(this.keyFetcher),
+      { algorithms: [ALGORITHM_RS256] })
+      .then(isValid => {
+        return isValid ? Promise.resolve() : Promise.reject(error);
+      });
+  }
+}
+
+function getKeyCallback(fetcher: KeyFetcher): jwt.GetPublicKeyOrSecret {
+  return (header: jwt.JwtHeader, callback: jwt.SigningKeyCallback) => {
+    const kid = header.kid || '';
+    fetcher.fetchPublicKeys().then((publicKeys) => {
+      if (!Object.prototype.hasOwnProperty.call(publicKeys, kid)) {
+        callback(new Error(NO_MATCHING_KID_ERROR_MESSAGE));
+      } else {
+        callback(null, publicKeys[kid]);
+      }
+    })
+      .catch(error => {
+        callback(error);
+      });
+  }
+}
+
+export class EmulatorSignatureVerifier implements SignatureVerifier {
+  public verify(token: string): Promise<void> {
+    const error = new JwtError(JwtErrorCode.INVALID_TOKEN,
+      'The provided token has invalid signature.');
+
+    // Signature checks skipped for emulator; no need to fetch public keys.
+    return isSignatureValid(token, '')
+      .then(isValid => {
+        return isValid ? Promise.resolve() : Promise.reject(error);
+      });
+  }
+}
+
+function isSignatureValid(token: string, secretOrPublicKey: jwt.Secret | jwt.GetPublicKeyOrSecret,
+  options?: jwt.VerifyOptions): Promise<boolean> {
+  return new Promise((resolve, reject) => {
+    jwt.verify(token, secretOrPublicKey, options,
+      (error: jwt.VerifyErrors | null) => {
+        if (!error) {
+          return resolve(true);
+        }
+        if (error.name === 'TokenExpiredError') {
+          return reject(new JwtError(JwtErrorCode.TOKEN_EXPIRED,
+            'The provided token has expired. Get a fresh token from your ' +
+            'client app and try again.'));
+        } else if (error.name === 'JsonWebTokenError') {
+          const prefix = 'error in secret or public key callback: ';
```

Comment:
Should probably define a constant for this too: `JWT_CALLBACK_ERROR_PREFIX` or similar. And add a comment indicating how it came to be.